src/main/webapp/app/exam/manage/student-exams/student-exam-timeline/student-exam-timeline.component.ts

```diff
@@ -0,0 +1,388 @@
+            return this.datePipe.transform(this.options.stepsArray?.at(value)?.value, 'time', true);
+        };
+        this.options = newOptions;
+    }
+
+    /**
+     * Checks if the submission is a submission version
+     * @param object the object to check
+     */
+    isSubmissionVersion(object: SubmissionVersion | Submission | undefined) {
+        if (!object) {
+            return false;
+        }
+        const submissionVersion = object as SubmissionVersion;
+        return submissionVersion.id && submissionVersion.createdDate && submissionVersion.content && submissionVersion.submission;
+    }
+
+    /**
+     * Retrieve all submission versions/submissions for all exercises of the exam
+     */
+    retrieveSubmissionDataAndTimeStamps() {
+        const submissionObservables: Observable<SubmissionVersion[] | Submission[]>[] = [];
+        this.studentExam.exercises?.forEach((exercise) => {
+            if (exercise.type === ExerciseType.PROGRAMMING) {
+                const id = exercise.studentParticipations![0].id!;
+                const programmingSubmission = this.submissionService.findAllSubmissionsOfParticipation(id).pipe(map(({ body }) => body!));
+                submissionObservables.push(programmingSubmission);
+            } else if (exercise.type === ExerciseType.FILE_UPLOAD) {
+                const id = exercise.studentParticipations![0].id!;
+                submissionObservables.push(this.submissionService.findAllSubmissionsOfParticipation(id).pipe(map(({ body }) => body!)));
+            } else {
+                submissionObservables.push(
+                    this.submissionVersionService.findAllSubmissionVersionsOfSubmission(exercise.studentParticipations![0].submissions![0].id!).pipe(
+                        mergeMap((versions) => versions),
+                        toArray(),
+                    ),
+                );
+            }
+        });
+        return forkJoin(...submissionObservables);
+    }
+
+    /**
+     * Sorts the time stamps in ascending order
+     */
+    private sortTimeStamps() {
+        this.submissionTimeStamps = this.submissionTimeStamps.sort((date1, date2) => (date1.isAfter(date2) ? 1 : -1));
+    }
+
+    /**
+     * This method is called when the user clicks on the next or previous button in the navigation bar or on the slider.
+     * @param exerciseChange contains the exercise to which the user wants to navigate to and the submission that should be displayed
+     */
+    onPageChange(exerciseChange: {
+        overViewChange: boolean;
+        exercise?: Exercise;
+        forceSave: boolean;
+        submission?: ProgrammingSubmission | SubmissionVersion | FileUploadSubmission;
+    }): void {
+        const activeComponent = this.activePageComponent;
+        if (activeComponent) {
+            activeComponent.onDeactivate();
+        }
+
+        if (!exerciseChange.submission) {
+            // only change the submission if the exercise has changed
```

Comment:
For which types of exercises can this be the case?

Reply:
for all exercises. What we preven with this check,, is updating the view if you press on the same button multiple times in the navigation bar.

Reply:
Ahh, I see 👍